File: server/excalibur_server/src/crypto/elliptic/elgamal.py
from excalibur_server.src.crypto.elliptic.ristretto255 import Ristretto255


class ElGamal:
    @staticmethod
    def encrypt(public_key: Ristretto255, m: Ristretto255, blind_scalar: int | None = None) -> bytes:
        y = blind_scalar or Ristretto255.random_scalar()

        s = public_key * y
        c1 = y * Ristretto255.GENERATOR
        c2 = m + s

        return c1.to_bytes() + c2.to_bytes()

    @staticmethod
    def decrypt(private_key: int, ciphertext: bytes) -> Ristretto255:
        c1 = Ristretto255.from_bytes(ciphertext[: Ristretto255.KEY_LENGTH])
        c2 = Ristretto255.from_bytes(ciphertext[Ristretto255.KEY_LENGTH :])
        s = c1 * private_key
        m = c2 - s
        return m

File: app/packages/main/src/lib/crypto/elliptic/elgamal.ts
import Ristretto255 from "./ristretto255";

export default class ElGamal {
    static encrypt(publicKey: Ristretto255, m: Ristretto255, blindScalar?: bigint): Buffer {
        const y = blindScalar ?? Ristretto255.randomScalar();

        const s = publicKey.mul(y);
        const c1 = Ristretto255.GENERATOR.mul(y);
        const c2 = m.add(s);

        return Buffer.concat([c1.toBytes(), c2.toBytes()]);
    }

    static decrypt(privateKey: bigint, ciphertext: Buffer): Ristretto255 {
        const c1 = Ristretto255.fromBytes(ciphertext.subarray(0, Ristretto255.KEY_LENGTH));
        const c2 = Ristretto255.fromBytes(ciphertext.subarray(Ristretto255.KEY_LENGTH));

        const s = c1.mul(privateKey);
        const m = c2.sub(s);

        return m;
    }
}
